### PR TITLE
OCPCLOUD-2137: Check scale from zero annotations on MachineSets

### PR DESCRIPTION
**Pull request by odvarkadaniel**
Currently, we support only `machine.openshift.io` annotations for scaling from zero - this PR introduced logic that let's the user use the upstream prefix `capacity.cluster-autoscaler.kubernetes.io` for this purpose. We also need to make sure to correctly calculate between `memoryMb` and `memory` suffixes for their corresponding prefixes. We want to have both sets of these annotation present on the MachineSet.

### CODE REVIEW

**Comment by odvarkadaniel** [comment]
cc @elmiko for review

File: pkg/controller/machineautoscaler/machinetarget.go [hunk 1]
```diff
@@ -280,3 +291,67 @@ func (mt *MachineTarget) WarningForInvalidGPUAcceleratorLabel() string {
...
+	annotations := mt.GetAnnotations()
+
+	if annotations == nil {
+		return nil
+	}
+
+	err := checkScaleFromZeroAnnotations(annotations)
+	if err != nil {
+		// canScale, err = checkNewScaleFromZeroAnnotations(annotations)
+		// if !canScale && err != nil {
```
Comment (elmiko) [inline pkg/controller/machineautoscaler/machinetarget.go:309]: are these comments left over?

File: pkg/controller/machineautoscaler/machinetarget.go [hunk 2]
```diff
@@ -280,3 +291,67 @@ func (mt *MachineTarget) WarningForInvalidGPUAcceleratorLabel() string {
...
+			return fmt.Errorf("could not parse integer: %w", err)
+		}
+		annotations[memoryKey] = resource.NewQuantity(memInt*util.MiB, resource.DecimalSI).String()
+	}
+
+	// Original GPU annotation is split into 2 upstream annotations
+	gpu := parseKey(annotations, gpuCountKeyOriginal)
+	if gpu != "" {
+		annotations[gpuCountKey] = gpu
+		annotations[gpuTypeKey] = "nvidia.com/gpu"
```
Comment (elmiko) [inline pkg/controller/machineautoscaler/machinetarget.go:340]: could you add a `// TODO ` comment here saying that we need to come back and fix this once we have the proper gpu types.

**Review by elmiko — COMMENTED** [review]
this is looking good Daniel, i like the comments too. just a couple comments.

File: pkg/controller/machineautoscaler/machinetarget.go (hunk 2 above)
Comment (odvarkadaniel) [reply pkg/controller/machineautoscaler/machinetarget.go:339]: Added.

**Comment by odvarkadaniel** [comment]
> this is looking good Daniel, i like the comments too. just a couple comments.

Addressed your comments, Mike. Should be ready to review.

**Comment by odvarkadaniel** [comment]
The CI is now expected to fail until the module work is merged in MAO (https://github.com/openshift/machine-api-operator/pull/1169).

File: pkg/controller/machineautoscaler/machinetarget.go [hunk 3]
```diff
@@ -280,3 +281,65 @@ func (mt *MachineTarget) WarningForInvalidGPUAcceleratorLabel() string {
...
+	// Deprecated GPU annotation is split into 2 upstream annotations.
+	// TODO: Once we introduce proper gpu types, the gpu type in the annotation module in MAO needs to be changed.
+	gpu, err := annotationsutil.ParseMachineSetAnnotationKey(annotations, annotationsutil.GpuCountKeyDeprecated)
+	if err != nil {
+		return fmt.Errorf("could not parse key %s: %w", annotationsutil.GpuCountKeyDeprecated, err)
+	}
+	annotations = annotationsutil.SetGpuCountAnnotation(annotations, gpu)
+	// TODO: Once we introduce proper gpu types, the function SetGpuTypeAnnotation
+	// should take in the correct value as a second argument.
+	annotations = annotationsutil.SetGpuTypeAnnotation(annotations, "")
```
Comment (elmiko) [inline pkg/controller/machineautoscaler/machinetarget.go:336]: i know we don't do anything with the second value, but let's pass  the nvidia value anyways just to make it clear what is happening.

we probably should not have defaulted the library function, but i think this is a minor detail.

**Review by elmiko — COMMENTED** [review]
this is looking good to me, i just have one minor request

File: pkg/controller/machineautoscaler/machinetarget.go (hunk 3 above)
Comment (odvarkadaniel) [reply pkg/controller/machineautoscaler/machinetarget.go:336]: Agree. Changing the value we pass into the function.

**Comment by odvarkadaniel** [comment]
> this is looking good to me, i just have one minor request

Fixed. Should be good to go.

**Pull request merged**